cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -15,46 +15,368 @@
  */
 #pragma once
 
+#include <prims/extract_transform_v_frontier_outgoing_e.cuh>
+#include <prims/per_v_transform_reduce_incoming_outgoing_e.cuh>
+#include <prims/transform_reduce_v_frontier_outgoing_e_by_dst.cuh>
+#include <prims/update_edge_src_dst_property.cuh>
+#include <prims/update_v_frontier.cuh>
+#include <prims/vertex_frontier.cuh>
+
 #include <cugraph/algorithms.hpp>
+#include <cugraph/detail/utility_wrappers.hpp>
+#include <cugraph/edge_src_dst_property.hpp>
 #include <cugraph/utilities/error.hpp>
+#include <cugraph/vertex_partition_device_view.cuh>
+
+#include <thrust/functional.h>
+#include <thrust/reduce.h>
 
 #include <raft/core/handle.hpp>
 
+//
+// The formula for BC(v) is the sum over all (s,t) where s != v != t of
+// sigma_st(v) / sigma_st.  Sigma_st(v) is the number of shortest paths
+// that pass through vertex v, whereas sigma_st is the total number of shortest
+// paths.
+namespace {
+
+template <typename vertex_t>
+struct brandes_e_op_t {
+  const vertex_t max_vertex_t{std::numeric_limits<vertex_t>::max()};
+
+  template <typename value_t, typename ignore_t>
+  __device__ thrust::optional<value_t> operator()(
+    vertex_t, vertex_t, value_t src_property, vertex_t dst_property, ignore_t) const
+  {
+    thrust::optional<value_t> result{thrust::nullopt};
+    return (dst_property < max_vertex_t) ? result : thrust::make_optional(src_property);
+  }
+};
+
+}  // namespace
+
 namespace cugraph {
 namespace detail {
 
 template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>> brandes_bfs(
+  raft::handle_t const& handle,
+  graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+  std::optional<edge_property_view_t<edge_t, weight_t const*>> edge_weight_view,
+  vertex_frontier_t<vertex_t, void, multi_gpu, true>& vertex_frontier,
+  bool do_expensive_check)
+{
+  //
+  // Do BFS with a multi-output.  If we're on hop k and multiple vertices arrive at vertex v,
+  // add all predecessors to the predecessor list, don't just arbitrarily pick one.
+  //
+  // Predecessors could be a CSR if that's helpful for doing the backwards tracing
+  constexpr vertex_t invalid_distance = std::numeric_limits<vertex_t>::max();
+  constexpr int bucket_idx_cur{0};
+  constexpr int bucket_idx_next{1};
+
+  rmm::device_uvector<vertex_t> sigma(graph_view.local_vertex_partition_range_size(),
+                                      handle.get_stream());
```

Comment:
I will need to spend some time in the code to ponder the rest of your comments in this file.  This might need to be an `edge_t`, I'll work through the code and verify.

Reply:
``` there will be at most number_of_vertices() paths```=>Is this true with multi-edges as well?

What happens if there are two vertices 0 and 1 and trillions of multi-edges between vertex 0 & 1?